# Using Threads to Run Code Simultaneously
使用线程同时运行代码

In most current operating systems, an executed program’s code is run in a process, and the operating system will manage multiple processes at once. Within a program, you can also have independent parts that run simultaneously. The features that run these independent parts are called threads. For example, a web server could have multiple threads so that it could respond to more than one request at the same time.
在当前的大多数操作系统中，一个已执行程序的代码运行在一个进程中，操作系统将同时管理多个进程。在一个程序中，也可以有同时运行的独立部分。运行这些独立部分的特性称为线程。例如，一个web服务器可以有多个线程，这样它就可以同时响应多个请求。

Splitting the computation in your program into multiple threads to run multiple tasks at the same time can improve performance, but it also adds complexity. Because threads can run simultaneously, there’s no inherent guarantee about the order in which parts of your code on different threads will run. This can lead to problems, such as:
将程序中的计算分成多个线程来同时运行多个任务可以提高性能，但也增加了复杂性。
因为线程可以同时运行，所以对于不同线程上代码的运行顺序没有内在的保证。这可能会导致一些问题，例如:

Race conditions, where threads are accessing data or resources in an inconsistent order
Deadlocks, where two threads are waiting for each other, preventing both threads from continuing
Bugs that happen only in certain situations and are hard to reproduce and fix reliably
Rust attempts to mitigate the negative effects of using threads, but programming in a multithreaded context still takes careful thought and requires a code structure that is different from that in programs running in a single thread.
竞争条件，线程以不一致的顺序访问数据或资源 
死锁，两个线程相互等待，阻止两个线程继续 
只有在某些情况下才会发生的错误，并且很难重现和可靠地修复 
Rust试图减轻使用线程的负面影响，但是在多线程上下文中编程仍然需要仔细考虑，并且需要不同于在单个线程中运行的程序的代码结构。

Programming languages implement threads in a few different ways, and many operating systems provide an API the language can call for creating new threads. The Rust standard library uses a 1:1 model of thread implementation, whereby a program uses one operating system thread per one language thread. There are crates that implement other models of threading that make different tradeoffs to the 1:1 model.
编程语言以几种不同的方式实现线程，许多操作系统提供了语言可以调用的创建新线程的API。
Rust标准库使用线程实现的1:1模型，即程序对每个语言线程使用一个操作系统线程。
有一些crate实现了其他线程模型，它们对1:1模型做出了不同的权衡。

## Creating a New Thread with spawn
用spawn创建一个新线程

To create a new thread, we call the thread::spawn function and pass it a closure (we talked about closures in Chapter 13) containing the code we want to run in the new thread. The example in Listing 16-1 prints some text from a main thread and other text from a new thread:
要创建新线程，我们调用thread::spawn函数，并向它传递一个闭包(我们在第13章讨论过闭包)，其中包含我们希望在新线程中运行的代码。代码清单16-1中的示例从主线程打印一些文本，从新线程打印另一些文本:

Filename: src/main.rs
``` rust
use std::thread;
use std::time::Duration;

fn main() {
    thread::spawn(|| {
        for i in 1..10 {
            println!("hi number {} from the spawned thread!", i);
            thread::sleep(Duration::from_millis(1));
        }
    });

    for i in 1..5 {
        println!("hi number {} from the main thread!", i);
        thread::sleep(Duration::from_millis(1));
    }
}
```
Listing 16-1: Creating a new thread to print one thing while the main thread prints something else
清单16-1:创建一个新线程打印一件东西，而主线程打印另一件东西

Note that when the main thread of a Rust program completes, all spawned threads are shut down, whether or not they have finished running. The output from this program might be a little different every time, but it will look similar to the following:
请注意，当Rust程序的主线程完成时，所有派生线程都将关闭，无论它们是否已经完成运行。这个程序的输出可能每次都略有不同，但看起来与下面类似:

```
hi number 1 from the main thread!
hi number 1 from the spawned thread!
hi number 2 from the main thread!
hi number 2 from the spawned thread!
hi number 3 from the main thread!
hi number 3 from the spawned thread!
hi number 4 from the main thread!
hi number 4 from the spawned thread!
hi number 5 from the spawned thread!
```

The calls to thread::sleep force a thread to stop its execution for a short duration, allowing a different thread to run. The threads will probably take turns, but that isn’t guaranteed: it depends on how your operating system schedules the threads. In this run, the main thread printed first, even though the print statement from the spawned thread appears first in the code. And even though we told the spawned thread to print until i is 9, it only got to 5 before the main thread shut down.
对thread::sleep的调用强制一个线程在短时间内停止执行，允许另一个线程运行。
线程可能会轮流执行，但这并不能保证:这取决于操作系统如何调度线程。在这次运行中，主线程首先打印，尽管衍生线程的print语句首先出现在代码中。尽管我们告诉派生线程要打印到i为9，但它只在主线程关闭前打印到5。

If you run this code and only see output from the main thread, or don’t see any overlap, try increasing the numbers in the ranges to create more opportunities for the operating system to switch between the threads.
如果运行这段代码，只看到主线程的输出，或者没有看到任何重叠的输出，请尝试增加范围内的数字，为操作系统在线程之间切换创造更多机会。

Waiting for All Threads to Finish Using join Handles
The code in Listing 16-1 not only stops the spawned thread prematurely most of the time due to the main thread ending, but because there is no guarantee on the order in which threads run, we also can’t guarantee that the spawned thread will get to run at all!
使用join句柄等待所有线程完成 
代码清单16-1中的代码不仅在大多数情况下会因为主线程结束而提前停止衍生线程，而且由于无法保证线程的运行顺序，我们也无法保证衍生线程能够运行!

We can fix the problem of the spawned thread not running or ending prematurely by saving the return value of thread::spawn in a variable. The return type of thread::spawn is JoinHandle. A JoinHandle is an owned value that, when we call the join method on it, will wait for its thread to finish. Listing 16-2 shows how to use the JoinHandle of the thread we created in Listing 16-1 and call join to make sure the spawned thread finishes before main exits:
我们可以通过将thread::spawn的返回值保存在一个变量中来修复派生线程不运行或过早结束的问题。thread::spawn的返回类型是JoinHandle。JoinHandle是一个拥有的值，当我们调用它的join方法时，它会等待它的线程完成。代码清单16-2展示了如何使用代码清单16-1中创建的线程的JoinHandle，并调用join来确保派生线程在主线程退出之前结束:

Filename: src/main.rs
``` rust
use std::thread;
use std::time::Duration;

fn main() {
    let handle = thread::spawn(|| {
        for i in 1..10 {
            println!("hi number {} from the spawned thread!", i);
            thread::sleep(Duration::from_millis(1));
        }
    });

    for i in 1..5 {
        println!("hi number {} from the main thread!", i);
        thread::sleep(Duration::from_millis(1));
    }

    handle.join().unwrap();
}
```
Listing 16-2: Saving a JoinHandle from thread::spawn to guarantee the thread is run to completion
清单16-2:保存一个JoinHandle from thread::spawn以保证线程运行到完成

Calling join on the handle blocks the thread currently running until the thread represented by the handle terminates. Blocking a thread means that thread is prevented from performing work or exiting. Because we’ve put the call to join after the main thread’s for loop, running Listing 16-2 should produce output similar to this:
在句柄上调用join会阻塞当前运行的线程，直到句柄所表示的线程终止。阻塞线程是指线程被阻止执行工作或退出。因为我们把join调用放在了主线程的for循环之后，运行代码清单16-2应该会得到类似下面的输出:

```
hi number 1 from the main thread!
hi number 2 from the main thread!
hi number 1 from the spawned thread!
hi number 3 from the main thread!
hi number 2 from the spawned thread!
hi number 4 from the main thread!
hi number 3 from the spawned thread!
hi number 4 from the spawned thread!
hi number 5 from the spawned thread!
hi number 6 from the spawned thread!
hi number 7 from the spawned thread!
hi number 8 from the spawned thread!
hi number 9 from the spawned thread!
```

The two threads continue alternating, but the main thread waits because of the call to handle.join() and does not end until the spawned thread is finished.
两个线程继续交替执行，但主线程会因为调用handle.join()而等待，直到衍生线程结束。

But let’s see what happens when we instead move handle.join() before the for loop in main, like this:
但是，让我们看看将handle.join()移动到main中的for循环之前会发生什么，如下所示:

Filename: src/main.rs
```
use std::thread;
use std::time::Duration;

fn main() {
    let handle = thread::spawn(|| {
        for i in 1..10 {
            println!("hi number {} from the spawned thread!", i);
            thread::sleep(Duration::from_millis(1));
        }
    });

    handle.join().unwrap();

    for i in 1..5 {
        println!("hi number {} from the main thread!", i);
        thread::sleep(Duration::from_millis(1));
    }
}
```

The main thread will wait for the spawned thread to finish and then run its for loop, so the output won’t be interleaved anymore, as shown here:
主线程将等待衍生线程完成，然后运行其for循环，因此输出将不再交错，如下所示:

```
hi number 1 from the spawned thread!
hi number 2 from the spawned thread!
hi number 3 from the spawned thread!
hi number 4 from the spawned thread!
hi number 5 from the spawned thread!
hi number 6 from the spawned thread!
hi number 7 from the spawned thread!
hi number 8 from the spawned thread!
hi number 9 from the spawned thread!
hi number 1 from the main thread!
hi number 2 from the main thread!
hi number 3 from the main thread!
hi number 4 from the main thread!
```

Small details, such as where join is called, can affect whether or not your threads run at the same time.
一些小细节，比如在哪里调用join，会影响线程是否同时运行。

Using move Closures with Threads
We'll often use the move keyword with closures passed to thread::spawn because the closure will then take ownership of the values it uses from the environment, thus transferring ownership of those values from one thread to another. In the “Capturing the Environment with Closures” section of Chapter 13, we discussed move in the context of closures. Now, we’ll concentrate more on the interaction between move and thread::spawn.
在线程中使用移动闭包 
我们经常使用move关键字来传递闭包给thread::spawn，因为闭包会从环境中获取它使用的值的所有权，从而将这些值的所有权从一个线程转移到另一个线程。
在第13章的“用闭包捕获环境”一节中，我们讨论了关于闭包的移动。
现在，我们将更多地关注move和thread::spawn之间的交互。

Notice in Listing 16-1 that the closure we pass to thread::spawn takes no arguments: we’re not using any data from the main thread in the spawned thread’s code. To use data from the main thread in the spawned thread, the spawned thread’s closure must capture the values it needs. Listing 16-3 shows an attempt to create a vector in the main thread and use it in the spawned thread. However, this won’t yet work, as you’ll see in a moment.
注意，在代码清单16-1中，传递给thread::spawn的闭包没有参数:在衍生线程的代码中，我们没有使用主线程的任何数据。
为了在衍生线程中使用主线程的数据，衍生线程的闭包必须捕获它需要的值。
代码清单16-3展示了如何在主线程中创建一个向量，并在衍生线程中使用它。然而，这还不起作用，稍后你就会看到。

Filename: src/main.rs

This code does not compile!
``` rust
use std::thread;

fn main() {
    let v = vec![1, 2, 3];

    let handle = thread::spawn(|| {
        println!("Here's a vector: {:?}", v);
    });

    handle.join().unwrap();
}
```
Listing 16-3: Attempting to use a vector created by the main thread in another thread
清单16-3:尝试在另一个线程中使用主线程创建的向量

The closure uses v, so it will capture v and make it part of the closure’s environment. Because thread::spawn runs this closure in a new thread, we should be able to access v inside that new thread. But when we compile this example, we get the following error:
这个闭包使用v，所以它会捕获v并使其成为闭包环境的一部分。
因为thread::spawn在新线程中运行这个闭包，所以我们应该能够在新线程中访问v。但是当我们编译这个例子时，我们得到了以下错误:

``` bash
$ cargo run
   Compiling threads v0.1.0 (file:///projects/threads)
error[E0373]: closure may outlive the current function, but it borrows `v`, which is owned by the current function
 --> src/main.rs:6:32
  |
6 |     let handle = thread::spawn(|| {
  |                                ^^ may outlive borrowed value `v`
7 |         println!("Here's a vector: {:?}", v);
  |                                           - `v` is borrowed here
  |
note: function requires argument type to outlive `'static`
 --> src/main.rs:6:18
  |
6 |       let handle = thread::spawn(|| {
  |  __________________^
7 | |         println!("Here's a vector: {:?}", v);
8 | |     });
  | |______^
help: to force the closure to take ownership of `v` (and any other referenced variables), use the `move` keyword
  |
6 |     let handle = thread::spawn(move || {
  |                                ++++
```

For more information about this error, try `rustc --explain E0373`.
error: could not compile `threads` due to previous error
Rust infers how to capture v, and because println! only needs a reference to v, the closure tries to borrow v. However, there’s a problem: Rust can’t tell how long the spawned thread will run, so it doesn’t know if the reference to v will always be valid.
有关此错误的更多信息，请尝试尝试`rustc——explain E0373`。 
错误:由于先前的错误，无法编译“线程” 
Rust推断如何捕获v，因为println!只需要对v的引用，闭包就会尝试借用v。
然而，有一个问题:Rust无法判断派生的线程将运行多长时间，因此它不知道对v的引用是否始终有效。

Listing 16-4 provides a scenario that’s more likely to have a reference to v that won’t be valid:
代码清单16-4提供了一种场景，其中对v的引用很可能是无效的:

Filename: src/main.rs

This code does not compile!
这段代码无法编译!
``` rust
use std::thread;

fn main() {
    let v = vec![1, 2, 3];

    let handle = thread::spawn(|| {
        println!("Here's a vector: {:?}", v);
    });

    drop(v); // oh no!

    handle.join().unwrap();
}
```
Listing 16-4: A thread with a closure that attempts to capture a reference to v from a main thread that drops v
清单16-4:一个带有闭包的线程，它试图从删除v的主线程捕获对v的引用

If Rust allowed us to run this code, there’s a possibility the spawned thread would be immediately put in the background without running at all. The spawned thread has a reference to v inside, but the main thread immediately drops v, using the drop function we discussed in Chapter 15. Then, when the spawned thread starts to execute, v is no longer valid, so a reference to it is also invalid. Oh no!
如果Rust允许我们运行此代码，则可能会立即将派生线程置于后台而根本不运行。
派生线程内部有一个v的引用，但主线程会立即使用第15章讨论过的drop函数删除v。
然后，当衍生的线程开始执行时，v不再有效，因此对它的引用也是无效的。噢,不!

To fix the compiler error in Listing 16-3, we can use the error message’s advice:
要修复清单16-3中的编译器错误，可以使用错误消息的提示信息:

help: to force the closure to take ownership of `v` (and any other referenced variables), use the `move` keyword
帮助:要强制闭包拥有`v`(以及其他被引用的变量)，使用`move`关键字
```
  |
6 |     let handle = thread::spawn(move || {
  |                                ++++
```
By adding the move keyword before the closure, we force the closure to take ownership of the values it’s using rather than allowing Rust to infer that it should borrow the values. The modification to Listing 
通过在闭包之前添加move关键字，我们强制闭包获取它正在使用的值的所有权，而不是允许Rust推断它应该借用这些值。对Listing的修改

16-3 shown in Listing 16-5 will compile and run as we intend:
代码清单16-5中的16-3将按照预期编译并运行:

Filename: src/main.rs
``` rust
use std::thread;

fn main() {
    let v = vec![1, 2, 3];

    let handle = thread::spawn(move || {
        println!("Here's a vector: {:?}", v);
    });

    handle.join().unwrap();
}
```
Listing 16-5: Using the move keyword to force a closure to take ownership of the values it uses
清单16-5:使用move关键字强制闭包获得它使用的值的所有权

We might be tempted to try the same thing to fix the code in Listing 16-4 where the main thread called drop by using a move closure. However, this fix will not work because what Listing 16-4 is trying to do is disallowed for a different reason. If we added move to the closure, we would move v into the closure’s environment, and we could no longer call drop on it in the main thread. We would get this compiler error instead:
我们可能会尝试做同样的事情来修复清单16-4中的代码，其中主线程通过使用move闭包调用drop。
不过，这个修复不会起作用，因为代码清单16-4试图做的事情是不允许的，原因与代码清单16-4不同。
如果我们给闭包添加move，我们就会把v移动到闭包的环境中，并且我们不能再在主线程中调用drop了。
我们会得到以下编译错误:

```
$ cargo run
   Compiling threads v0.1.0 (file:///projects/threads)
error[E0382]: use of moved value: `v`
  --> src/main.rs:10:10
   |
4  |     let v = vec![1, 2, 3];
   |         - move occurs because `v` has type `Vec<i32>`, which does not implement the `Copy` trait
5  | 
6  |     let handle = thread::spawn(move || {
   |                                ------- value moved into closure here
7  |         println!("Here's a vector: {:?}", v);
   |                                           - variable moved due to use in closure
...
10 |     drop(v); // oh no!
   |          ^ value used here after move
```

For more information about this error, try `rustc --explain E0382`.
error: could not compile `threads` due to previous error
Rust’s ownership rules have saved us again! We got an error from the code in Listing 16-3 because Rust was being conservative and only borrowing v for the thread, which meant the main thread could theoretically invalidate the spawned thread’s reference. By telling Rust to move ownership of v to the spawned thread, we’re guaranteeing Rust that the main thread won’t use v anymore. If we change Listing 16-4 in the same way, we’re then violating the ownership rules when we try to use v in the main thread. The move keyword overrides Rust’s conservative default of borrowing; it doesn’t let us violate the ownership rules.
有关此错误的更多信息，请尝试尝试`rustc——explain E0382`。 
错误:由于先前的错误，无法编译“线程” 
Rust的所有权规则再次拯救了我们!
我们从代码清单16-3中得到了一个错误，因为Rust是保守的，只在线程中借用了v，这意味着主线程理论上可以使派生线程的引用无效。
通过告诉Rust将v的所有权移动到衍生线程，我们保证Rust主线程将不再使用v。
如果我们以同样的方式修改代码清单16-4，那么当我们尝试在主线程中使用v时，就违反了所有权规则。
move关键字覆盖了Rust保守的借款违约;它不允许我们违反所有权规则。

With a basic understanding of threads and the thread API, let’s look at what we can do with threads.
有了对线程和线程API的基本了解，让我们看看我们可以用线程做什么。